### PR TITLE
fix(vwc-button): clicking a submit button results in generating a click on the first "submittable"

### PR DESCRIPTION
**Pull request by tveinfeld**
### Expected behaviour

When clicking a "submit" vwc-button, its "form ambassador" should be clicked to initiate form submission.

### Current behaviour

When clicking a "submit" vwc-button, the first "submittable" element of its hosting form is clicked.

### CODE REVIEW

**Comment by yinonov** [comment]
will [form data append](https://developer.mozilla.org/en-US/docs/Web/API/FormData/append) be satisfying enough to refactor the current approach?

**Comment by tveinfeld** [comment]
> will [form data append](https://developer.mozilla.org/en-US/docs/Web/API/FormData/append) be satisfying enough to refactor the current approach?

the formdata event carries no information regarding the "submitter" of a form (in case the form was submitted by an element), which is required

**Comment by yinonov** [comment]
> > will [form data append](https://developer.mozilla.org/en-US/docs/Web/API/FormData/append) be satisfying enough to refactor the current approach?
> 
> the formdata event carries no information regarding the "submitter" of a form (in case the form was submitted by an element), which is required

then we should delegate a value attribute as well, right?

File: components/button/src/vwc-button.ts [hunk 1]
```diff
@@ -88,6 +91,15 @@ export class VWCButton extends MWCButton {
...
+		[...changes.keys()]
+			.filter(attributeName => ['name', 'value'].includes(attributeName as string))
+			.forEach((attributeName) => {
+				this.#_hiddenButton.setAttribute(attributeName as string, (this as unknown as GenericMap)[attributeName as string]);
+			});
+	}
+
 	protected updated(changes: Map<string, boolean>): void {
 		if (changes.has('type')) {
 			this.#_hiddenButton?.setAttribute('type', this.getAttribute('type') ?? '');
```
Comment (YonatanKra) [inline components/button/src/vwc-button.ts:105]: I know it's not part of the PR, but if we refactor, we refactor :)
Why should `type` be handled here? Why can't it be in the `update` section?
Or why won't the others be in the `updated`? Just add `type` to the attributes array suggested above and delete the `updated`

**Review by YonatanKra — DISMISSED** [review]

File: components/button/src/vwc-button.ts [hunk 2]
```diff
@@ -88,6 +91,15 @@ export class VWCButton extends MWCButton {
 		}
 	}
 
+	protected override update(changes:PropertyValues):void {
```
[the remark this answers is not in the document]
Comment (tveinfeld) [reply components/button/src/vwc-button.ts:94]: it indicates that this method overrides a super's method

File: components/button/src/vwc-button.ts (hunk 1 above)
Comment (tveinfeld) [reply components/button/src/vwc-button.ts:107]: if it's all the same 2 u, let's do it on a different PR, as I'm not entirely familiar with the reasons that led to the current implementation 🙏

File: components/button/src/vwc-button.ts [hunk 3]
```diff
@@ -8,7 +8,7 @@ import { styles as mwcButtonStyles } from '@material/mwc-button/styles.css.js';
 import { style as styleCoupling } from '@vonage/vvd-style-coupling/mdc-vvd-coupling.css';
 import { Connotation, Layout, Shape } from '@vonage/vvd-foundation/constants';
 import { html, TemplateResult } from 'lit-element';
-import { requestSubmit } from '@vonage/vvd-foundation/form-association';
```
Comment (YonatanKra) [inline components/button/src/vwc-button.ts:11]: Is `requestSubmit` in usage anywhere now?

File: components/button/test/button.test.js [hunk 4]
```diff
@@ -38,7 +38,7 @@ describe('button', () => {
 		expect(b.shadowRoot.innerHTML).to.equalSnapshot();
 	});
 
-	describe('Form Association', function () {
+	describe('Form Association', async function () {
```
Comment (YonatanKra) [inline components/button/test/button.test.js:41]: Why do you need this to be `async`?

File: components/button/test/button.test.js [hunk 5]
```diff
@@ -47,16 +47,20 @@ describe('button', () => {
 			let submitted = false;
 			const addedElements = addElement(
 				textToDomToParent(
-					`<form onsubmit="return false" name="testForm" id="testForm"><${COMPONENT_NAME}>Button Text</${COMPONENT_NAME}></form>`
```
Comment (YonatanKra) [inline components/button/test/button.test.js:50]: Why did you change an existing test?

File: components/button/test/button.test.js [hunk 6]
```diff
@@ -47,16 +47,20 @@ describe('button', () => {
...
+					`<form onsubmit="return false" id="testForm"><${COMPONENT_NAME} name="testForm" value="testValue">Button Text</${COMPONENT_NAME}></form>`
 				)
 			);
 			await waitNextTask();
 			const formElement = addedElements[0];
 			const actualElement = formElement.firstChild;
-			formElement.addEventListener('submit', () => (submitted = true));
+			formElement.addEventListener('submit', ({ submitter }) => {
+				submitted =
```
Comment (YonatanKra) [inline components/button/test/button.test.js:57]: You could have done something like:
`expect(submitter === actualElement.querySelector('button')).to.equal(true)`;
This will also make sure they are the same, no?
All in all there are 3 things tested:
1. That the form is submitted
2. That if we add name and type they are reflected
3. That the submitter is the hidden button
These should be in 3 different `it`s.

Comment (tveinfeld) [reply components/button/test/button.test.js:56]: i was following your verbal recommendation :)

i would definitely prefer using expect in the way you offered. however, since i'm not the original author of this test, and it was already reviewed and approved this way, i'd rather not refactor anything.

1. if the form is submitted without the relevant name and value passed, there is no significance to the submission.
2. it is checked - both are set on 'vwc-button''s setup.
3. i'm not sure there's significance in verifying that the submitter is the hidden button - that's an implementation detail which might change (for instance - input instead of button etc.).

File: components/button/test/button.test.js (hunk 5 above)
Comment (tveinfeld) [reply components/button/test/button.test.js:49]: that's the default submission test.. there is no significance in testing submission without value & name alongside another test without them.

File: components/button/src/vwc-button.ts (hunk 1 above)
Comment (tveinfeld) [reply components/button/src/vwc-button.ts:110]: for me personally, my code is a bit more readable.. @yinonov - can you help us decide which way to go?

File: components/button/test/button.test.js (hunk 6 above)
Comment (yinonov) [reply components/button/test/button.test.js:56]: notice `submitter` compatibility. how will it pass in Safari?
image

Comment (tveinfeld) [reply components/button/test/button.test.js:56]: good catch! we need to rethink of ways to test this one then

**Comment by tveinfeld** [comment]
Hey all 👋
i've refactored the test following @yinonov's remark.. all previous comments are void, please re-process 🙇🏻

**Comment by yinonov** [comment]
> Hey all 👋
> 
> i've refactored the test following @yinonov's remark.. all previous comments are void, please re-process 🙇🏻
> 
> 
> 
> A bit about the new methodology:
> 
> 
> 
> i'm submitting a form into a blank iframe, then inspecting its 'window.location.search' to make sure it includes the required fields from the "name" and "value" attributes of "vwc-button"
> 
> 
> 
> Sorry for the inconvenience..

U mean u intercept the post back?

**Comment by tveinfeld** [comment]
> U mean u intercept the post back?

i'm taking a peek at the URL launched in the iframe to establish that it contains the relevant form data in its query string.

**Comment by yinonov** [comment]
> > U mean u intercept the post back?
> 
> i'm taking a peek at the URL launched in the iframe to establish that it contains the relevant form data in its query string.

Smart!
